### ISSUE TITLE
Render config with only one key when unresolved

### ISSUE THREAD

**Issue by hloos** [typesafehub/config#412]
Better to explain using an example :)

I have a origin config (called oldConfig) and a new one - which the difference between the oldConfig and the newOne - (called diff).

diff=

<pre>{
    "com" : {
        "test" : {
            "stringToMerge" : "newString",
            "stringWithTagToMerge" : ${com.hydrac.enable}"WithTag",
        }
    }
}</pre>

oldConfig=

<pre>{
    "com" : {
        "test" : {
            "stringToMerge" : "myTest",
            "stringWithTagToMerge" : ${com.hydrac.string}"WithTag",
        }
    }
}</pre>

diff.withFallBack(currentConfig)=
<pre>{
    "com" : {
        "test" : {
            "stringToMerge" : "newString",
                        "stringWithTagToMerge" : ${com.hydrac.string}"WithTag",
            "stringWithTagToMerge" : ${com.hydrac.enable}"WithTag"
        }
    }
}</pre>

In some cases, it should be useful to add an Option 'ignoreUnresolvedTag' into the com.typesafe.config.ConfigRenderOptions to fallback the unresolved values as simple values which will produce after rendering something like :

diff.withFallBack(currentConfig)=
<pre>{
    "com" : {
        "test" : {
            "stringToMerge" : "newString",
            "stringWithTagToMerge" : ${com.hydrac.enable}"WithTag"
        }
    }
}</pre>

**Comment by havocp** [comment]
I'm not sure I understand but maybe the issue here is in withFallback; it could drop an extra field that can't possibly be used? It may already be smart enough to do this if a field is overridden by a constant value so perhaps that check could instead be for whether a value can possibly become undefined?

**Comment by hloos** [comment]
After testing, it is indeed smart enough to render only the constant value instead of the previous placeholder.
The problem with checking if the value is possibly become undefined is the fact we do not know if, at the runtime, the placeholder will be replaceable or not.

**Comment by havocp** [comment]
We know that `${com.hydrac.enable}"WithTag"` can't be undefined... it could be an error, but not undefined. (Because of the `"WithTag"` constant part.) So `"stringWithSubstitutionToMerge" : ${com.hydrac.string}"WithTag"` should be possible to drop.

**Comment by hloos** [comment]
Unfortunately, I also have the case of something like: 
"substitutionToMerge" : ${com.test.string}

And I am in a packaging step, so with no clue on the running environment. That's why I need to ignore the undefined case.

**Comment by havocp** [comment]
The rendering code has no idea what's undefined; to determine that we would have to attempt to resolve substitutions and strip this out before we get to the render step. There is an "allow unresolved" option on the resolve() method, but it doesn't remove the unresolved nodes. Perhaps what you want is some sort of removeUnresolved() method?

Note that you can use `${?com.test.string}` to make a resolution optional, as well.

**Comment by hloos** [comment]
Thanks for the tip :)
However, here, the problem is not how to resolve it, but only how to render it without resolving it.

To my mind, since the rendering code has no idea what's undefined, it should render only the last value of a ConfigDelayedMerge instance (the value of ConfigA in our case).

configA.withFallback(configB) produces a Config instance containing:
```
"stringWithSubstitutionToMerge" -> ConfigDelayedMerge(${com.hydrac.string}"WithTag", ${com.hydrac.enable}"WithTag")
```

**Comment by havocp** [comment]
In many (most) cases it's important to render all the pieces, so people can see what their Config contains. The main point of rendering an unresolved config would be debugging, as far as I know.

What are you actually trying to do - what are you going to do with this rendered config?

**Comment by hloos** [comment]
My use-case: before deploying his application, the user want to modify the configuration.
Potentially, he may want to change the placeholder of a particular configuration.
Once changes are made, I need to merge the new config with the old one and write this result down in a configuration file.. He is expected knowing exactly what he is doing, so potential undefined substitution of placeholders are not an issue we need to take care.

Here, no running context, no way neither to guess the substitution corresponding values, so that's why I need to deal with placeholders without resolving them.

**Comment by havocp** [comment]
Maybe the issue is that overridden values with undefined substitutions are currently still an error? If we lazily resolve only values that will actually be used, then they would not be an error.

Dealing with unresolved values at render-time doesn't "make sense", it isn't an operation that can be defined in a coherent way I don't think. We need a fix that is defined in terms of the Config itself (prior to the render).

**Comment by hloos** [comment]
If such a use case does not make any sense at the render time, maybe a new method `Config#withFallback(ConfigMergeable, FallBackOptions)` could be added in `Config` with a new parameter allowing the user to specify how he want to treat the unresolved value. What do you think about that?

**Comment by havocp** [comment]
That is the right "place" in the API but I still don't know how to write a spec for the behavior you want. I don't feel like I understand the use case yet.

**Comment by hloos** [comment]
Just for information, the rendering for the use of the operator += (for a list) gives the same issue.

configA: `com.test.listToMerge+=12`
configB: `com.test.listToMerge+=16`

configA.withFallBack(configB):
<pre>{
    "com" : {
        "test" : {
                        "listToMerge" : ${?com.hydrac.listToMerge}[
                12
            ],
            "listToMerge" : ${?com.hydrac.listToMerge}[
                16
            ]
        }
    }
}</pre>

With your knowledge of typesafe.config and waiting for you to write a spec for this, do you see any clean workaround to allow me to render an unresolved config by "forcing" it to keep only one key-value pair?

Otherwise I can temporarily fork the project, implement my proposal and use this till an enhancement is done (even propose my implement with a pull-request)... I would like to avoid extracting key-value pairs, sort them before recreating a Config object (specially for list operator, it may be tricky)

**Comment by havocp** [comment]
I don't know a clean workaround. I feel like there is probably an easier and better way to arrange your big picture. Keeping exactly one unresolved value doesn't make any sense to me; it seems fragile and poorly-defined. Making a custom fork to get this behavior seems like a lot of work but of course you are welcome to do so.

What does make sense to me is that a value that wont be used (is overridden) could be allowed to be unresolved.

To be clear I'm not saying categorically that there's no change to the lib that makes sense here, but I don't know what it is yet.

The ways it's designed to be used:
* If you want to edit a single file use ConfigDocumentFactory and operate on the syntax tree
* But usually layers of customization are done by keeping separate files and merging them at runtime

If the problem is that an overridden substitution throws a "not resolved" error even though the value would not be used, I think we could simply fix that; we don't need to resolve a value that gets overridden. No API needed to fix this.

That is a coherent semantic (don't resolve what we don't need) while "discard all but one" feels like a broken heuristic that happens to work in certain cases.

But I'm not 100% confident thats the problem you are trying to solve.

**Comment by hloos** [comment]
I agree but the case I am presenting you here joins your first description: don't resolve what we don't need.

My user (a PS) prepares an application for a given customer before deploying it. One of the required step here is to customize the configuration. To do so, he uses his own packaging application which allows him to modify applications configuration and changes voluntary the value of `stringWithSubstitutionToMerge` from `${com.test.string}"WithTag"` to `${com.test.enable}"WithTag"`.

As you can see, `${com.test.string}"WithTag"` is not usefull anymore and won't be used, so it is useless to keep it after the merge.

Maybe the withFallBack method is finally not the appropriate method to do my merge since it seems to imply we are at the runtime and it is not really a fallback we are looking into here but more an explicit merge.

So, maybe an explicit ConfigMergeable#merge() method would be more appropriate which would strictly applicate the HOCON philosophy concerning the duplicate key `In HOCON, duplicate keys that appear later override those that appear earlier, unless both values are objects.` (coming from the doc https://github.com/typesafehub/config/blob/master/HOCON.md#duplicate-keys-and-object-merging)

**Comment by BrainStone** [comment]
I have an issue that is related this it seems:

Let's suppose I have these two configs:

`configA`
```
{
    "AccountDatabase" : {
        "database" : "Test",
        "user" : ${AccountDatabase.database}
    }
}
```

and a `configB`
```
{
    "AccountDatabase" : {
        "database" : "Test",
        "user" : "Foobar"
    }
}
```

then `configA.withFallback(configB)` results in:
```
{
    "AccountDatabase" : {
        "database" : "Test",
        # unresolved merge of 2 values follows (
        #     unmerged value 0 for key "user" from configB.conf: 4
        "user" : "Foobar",
        #     unmerged value 1 for key "user" from cinfigA: 4
        "user" : ${AccountDatabase.database}
        # ) end of unresolved merge

    }
}
```

Anyone can help with that? (Should I maybe make a new issue?)
